src/agent/application.cpp

```diff
@@ -99,6 +101,9 @@ void Application::Init(void)
 
 void Application::Deinit(void)
 {
+#if OTBR_ENABLE_TREL
+    otPlatTrelDisable(mNcp.GetInstance());
```

Comment:
Is otPlatTrelDisable already called in otSysDeinit -> platformDeinit -> platformTrelDeinit ?

Reply:
When doing `platformTrelDeinit`, Trel will be accessing its `mPublisher`, which is a reference to the border agent's `Publisher`. The border agent's publisher has already been destructed at that timepoint. Therefore, I need to explicitly disable Trel before the publisher is destructed.

Reply:
This is a little messy, but not because of this PR.
We have the plan to move `mPublisher` out of Application so that mPublisher is always available during the OT instance lifecycle. 

I think we should add a comment and TODO so that we can refactor it later.

Reply:
> The border agent's publisher has already been destructed at that timepoint. Therefore, I need to explicitly disable Trel before the publisher is destructed.

Why is the publisher already distructed? I think the Application is not destructed when calling `Deinit` ?

Reply:
Correction: Publisher is not destructed but all subscribed services are cleared in `Publisher::Stop()`. In `platformTrelDeinit()`, it will unsubscribe the TREL service. However, it cannot find the subscription so that it will crash at `assert(it != mSubscribedServices.end());`.

Reply:
IIUC, the issue is that `UnpublishService` is called after the Publisher is already stopped.

What about quiting `UnpublishService` if Publisher is already stopped.

It seems still logical for TrenDnssd to call on a stopped Publisher.

Reply:
Should we also verify whether Publisher is running on other Publisher methods?

Reply:
The Publisher must be running for other publisher method ?

I think we are already checking that the Publisher is running in other methods, like in `PublisherAvahi::PublishService`:
```
    VerifyOrExit(mState == State::kReady, error = OTBR_ERROR_INVALID_STATE);
    VerifyOrExit(mClient != nullptr, error = OTBR_ERROR_INVALID_STATE);
```

Should be fine to just fix UnpublishXxxx methods for now ?

Reply:
Done